contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,300 @@
+
+Another service we use is Redis to cache data and back the queue for apps. Many parts of the app are reusable and don’t need to be rebuilt each time we run the app. There is data we only need to keep around temporarily. Using Redis to cache data like this dramatically improved app performance and lowered the number of app servers we needed to run.
+
+We also use a couple of Redis patterns. First, we use [Redlock](https://redis.io/docs/manual/patterns/distributed-locks/) for the scheduler and job queue consumer. Redlock ensures these processes are only run on one machine because we only want them to run once. If they ran more than once, we’d have duplication and issues. For example, we only need one scheduler to do the basic work of saying what to do at what time. Second, we use [Pub/Sub](https://redis.io/docs/manual/pubsub/) to check for updates to the apps.
+
+We chose these because we used them elsewhere and they have a proven track record of working well at scale. They help the app server integrate with our existing infrastructure.
+
+### Integrating into the plugin server
+
+To ensure apps connected to the rest of the ingestion pipeline, we decided to integrate apps into our plugin server. The plugin server runs processes for ingesting, formatting, and writing data from users to storage. It runs on Node (because of the decision we made in the hackathon about speed and VMs).
+
+To show you how we did that, here’s a complete overview of what the plugin server looks like (our app server only utilizes some parts):
+
+![Plugin server](../images/blog/how-we-built-an-app-server/plugin-server.png)
+
+The main thread routes incoming tasks to the right location. It receives data and starts threads to complete them. We decided to use [Piscina](https://github.com/piscinajs/piscina) (a node.js worker pool) to create and manage the threads. We choose Piscina because it abstracts away a lot of the management of threads we would need to do to scale. The main thread also handles the functionality of scheduling and job queuing. The result creates tasks and sends them to worker threads to complete.
+
+### Doing the work (in worker threads)
+
+Worker threads receive tasks from the main thread and execute them. Some of the tasks, like `processEvent` or `runEveryMinute` use the callable functions (app code) we detailed above. Worker threads contain the VMs, as well as the ingestion logic and connections to extensions and libraries. Each worker thread can run up to 10 tasks at the same time.
+
+![Worker thread](../images/blog/how-we-built-an-app-server/worker-thread.png)
+
+When the worker thread finishes its task, it returns to the main thread for further processing. This could include more modification through apps, PostHog person processing, or writing to ClickHouse. A final `onEvent` task runs once all the processing completes, which is useful for functions like exporting or alerting.
```

Comment:
The worker thread doesn't return to the main thread, the worker thread remains where it is. It will just get another new event from the main thread, when it has something to send.

> This could include more modification through apps, PostHog person processing, or writing to ClickHouse.

"more modification through apps" -- more modification of what? what is "person processing"?

Just say "takes the next event to process or other scheduled task"